query.go

```diff
@@ -778,120 +772,272 @@ func (s *ChainService) GetCFilter(blockHash chainhash.Hash,
 		return nil, err
 	}
 
-	// We didn't get the filter from the DB, so we'll set it to nil and try
-	// to get it from the network.
-	filter = nil
+	// We didn't get the filter from the DB, so we'll try to get it from
+	// the network.
+	qo := defaultQueryOptions()
+	qo.applyQueryOptions(options...)
+
+	// If the query specifies an optimistic batch we will attempt to fetch
+	// the maximum number of filters in anticipation of calls for the
+	// following filters.
+	var maxFilters = uint32(1)
+	if qo.optimisticBatch {
+		maxFilters = wire.MaxGetCFiltersReqRange
+
+		// We need a longer timeout, since we are going to receive more
+		// than a single response.
+		options = append(options, Timeout(QueryBatchTimeout))
+	}
 
-	// In order to verify the authenticity of the filter, we'll fetch the
-	// target block header so we can retrieve the hash of the prior block,
-	// which is required to fetch the filter header for that block.
-	block, height, err := s.BlockHeaders.FetchHeader(&blockHash)
+	_, startHeight, err := s.BlockHeaders.FetchHeader(&blockHash)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("unable to get header for start "+
+			"block=%v: %v", blockHash, err)
 	}
-	if block.BlockHash() != blockHash {
-		return nil, fmt.Errorf("Couldn't get header for block %s "+
-			"from database", blockHash)
+
+	bestBlock, err := s.BestBlock()
+	if err != nil {
+		return nil, fmt.Errorf("unable to get best block: %v", err)
 	}
+	bestHeight := uint32(bestBlock.Height)
 
-	log.Debugf("Fetching filter for height=%v, hash=%v", height, blockHash)
+	// If the stop height with the maximum batch size is above our best
+	// known block, then we use the best block height instead.
+	stopHeight := startHeight + maxFilters - 1
+	if stopHeight > bestHeight {
+		stopHeight = bestHeight
+	}
 
-	// In addition to fetching the block header, we'll fetch the filter
-	// headers (for this particular filter type) from the database. These
-	// are required in order to verify the authenticity of the filter.
-	curHeader, err := getHeader(&blockHash)
+	stopHash, err := s.GetBlockHash(int64(stopHeight))
 	if err != nil {
-		return nil, fmt.Errorf("Couldn't get cfheader for block %s "+
-			"from database", blockHash)
+		return nil, fmt.Errorf("unable to get hash for "+
+			"stopHeight=%d: %v", stopHeight, err)
 	}
-	prevHeader, err := getHeader(&block.PrevBlock)
+
+	// In order to verify the authenticity of the received filters, we'll
```

Comment:
I'd say this logic along with the height index itself could be extracted to a method that returns the set of materials needed to generate the queries as well as validate the received filter headers, returning a struct of all the necessary items.

Reply:
Done.